### ISSUE TITLE
feat: asset support

### ISSUE THREAD

**Issue by eseidel** [shorebirdtech/shorebird#318]
What this would mean is making it possible to add images/assets via a patch.  Right now we're only able to change code, but adding asset support would mean that we would diff the assets as well (between the patch and the release) and send any changed assets (images, icons, etc.) with the patch to the devices in the field.

This requires some infrastructure work, mostly in preparing for now arbitrary-sized patches, which means we need to improve our download mechanisms on device (which are currently very simple because code-only patches are generally tiny).

As of Jan 2025, this is very close to the top of our list, but still not scheduled.

**Comment by eseidel** [comment]
Update:  So we built some of the infrastructure for this already.  We do store AABs from releases and we will check and report when you have changes other than dart-code in a patch.

What's missing to make this feature work is some changes to updater library:
https://github.com/shorebirdtech/updater/tree/main/library
to download the bundled assets and place them in a new directory and changes to the Flutter Engine code to create a DirectoryAssetBundle:
https://github.com/flutter/engine/blob/e4873bbb44f146329aa1737c762f7cdf129ec524/assets/directory_asset_bundle.h
and insert it into the AssetResolver chain.

There is work to figure out if you can add/modify and delete (not sure if delete is necessary) assets between patches, and if we can be smart about not re-downloading assets which aren't in the base APK, but were updated in patch 1, thus don't need to be downloaded again in patch 2, just copied over from patch 1's install directory (can be a later feature).

**Comment by eseidel** [comment]
OK, I'm starting this.

Notes:
* First add the directory creation in rust code.  Fill it with some random image from the internet.
* Build an example app with a different picture, same name, verify it shows the unpatched one.
* Send a patch to that app (which changes to the patched image name).  Verify that the image fails to load.
* Expose the assets directory path from rust to C++.
* In C++, create a DirectoryAssetBundle? pointing to the directory path from rust and PushFront onto the AssetManager.

Questions:
* Where do we actually insert the DirectoryAssetBundle?  Settings.assets_path seems to be an embedder api concept.  RunConfiguration seems to be partially integrated, but not always?
* https://github.com/shorebirdtech/engine/blob/655974aed33b608d1075cffae0fcaf06772d90d8/shell/platform/android/android_shell_holder.cc#L334 might be the location for Android?
* https://github.com/shorebirdtech/engine/blob/655974aed33b608d1075cffae0fcaf06772d90d8/shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm#L385 might be the iOS hook?
* Should we use DirectoryAssetBundle?  We could make one which reads out of a Zip like ApkAssetBundle does.
* Is UpdateResolverByType going to break us?  Is that used in release mode?

**Comment by eseidel** [comment]
UpdateAssetResolverByType appears to only be used to update the APK asset resolver, so we should be fine.

**Comment by eseidel** [comment]
Status update: I looked at this extensively but got drawn into other things.  I still don't think this will be very hard (and honestly is getting easier every day as we continue to build out our build and test infrastructure).

We're just focused on getting a production-ready (fast!) iOS shipped first, so this is still probably a few weeks away from getting our full concentration.

**Comment by eseidel** [comment]
One other note is that assets support likely has downstream effects, e.g.:

* When we add asset support we give developers the ability to make patches *a lot* bigger (by adding/updating large assets)
* will us to want to update our patch install logic.  Right now our patch install logic does not restrict itself to only unmetered networks (it should), but it also doesn't matter for customers much since patches are tiny (bytes or kilobytes).  Once patches can easily end up as megabytes (with asset support) we will want to make that more clear to developers and update our patching logic with partial downloads, etc.
* This will also change incentives pricing around patching (not immediately, but it will change things).  Right now we don't charge for network or disk (or have limits on either), since we're almost 100% responsible for patch size (and we have work we can do to make patches smaller).  

None of that needs to be solved for our v1 of assets support.  Just noting that asset support will have follow-on effects (that I'm excited to solve) and is just one more reason to put it *after* we get iOS into production (since that's by-far our top ask from customers).

**Comment by GeylanKalafMohe** [comment]
> One other note is that assets support likely has downstream effects, e.g.:
> 
> 
> 
> * When we add asset support we give developers the ability to make patches *a lot* bigger (by adding/updating large assets)
> 
> * will us to want to update our patch install logic.  Right now our patch install logic does not restrict itself to only unmetered networks (it should), but it also doesn't matter for customers much since patches are tiny (bytes or kilobytes).  Once patches can easily end up as megabytes (with asset support) we will want to make that more clear to developers and update our patching logic with partial downloads, etc.
> 
> * This will also change incentives pricing around patching (not immediately, but it will change things).  Right now we don't charge for network or disk (or have limits on either), since we're almost 100% responsible for patch size (and we have work we can do to make patches smaller).  
> 
> 
> 
> None of that needs to be solved for our v1 of assets support.  Just noting that asset support will have follow-on effects (that I'm excited to solve) and is just one more reason to put it *after* we get iOS into production (since that's by-far our top ask from customers). 

I wish to have a pricing guarantee. How can we as a company rely that the current tier subscription will be the same in upcoming Shorebird release cause of changes like assets charges...

**Comment by eseidel** [comment]
> I wish to have a pricing guarantee. How can we as a company rely that the current tier subscription will be the same in upcoming Shorebird release cause of changes like assets charges...

Our current pricing is honestly just a guess based on our underlying costs.  We essentially try to predict our costs and charge exactly that assuming full-usage of a plan.  We don't yet know if it's way too high or way too low yet.  We plan to eventually offer annual pricing: https://github.com/shorebirdtech/shorebird/issues/733, which would be a path towards a "guarantee".  We've changed our pricing model at least 3 times in the last 6 months (each time reducing our prices).  We're just still trying to figure out our costs as we continue to scale.  Regardless, happy to work with you on pricing that works for your team, just reach out to me at contact@shorebird.dev.  Thanks!

**Comment by GeylanKalafMohe** [comment]
Any plan or update when it's going to be implemented?

**Comment by eseidel** [comment]
This is entirely blocked behind us getting our 1.0 release out, which is blocked on finishing iOS: https://github.com/shorebirdtech/shorebird/issues/674.  I think it will depend on how quickly we can resolve that (we've been working on it for months, but I'm hopeful we're only a week or so away).  And then how much other customer work falls out of our 1.0 launch.  But asset support is pretty high on our list.  We already did some of the work needed, just need to go back and finish it.  Like Flutter, we take 👍 into account in prioritizing issues, so that helps.

**Comment by tabibo** [comment]
Hello @eseidel , Is there any updates regarding the asset support in the new patches?

**Comment by eseidel** [comment]
No updates.  We continue to focus all our efforts on https://github.com/shorebirdtech/shorebird/issues/674

**Comment by ronaldoandjames** [comment]
as the iOS 1.0 is about to release, is there a  schedule  to do this assets support?  it's a  important function 🥲

 thanks.

**Comment by robsombarros** [comment]
> No updates. We continue to focus all our efforts on #674

The issue #674 is closed. We need this asset update feature to enable the use of Shorebird. Do you have any estimate for fixing the 'missing assets' warning?

**Comment by eseidel** [comment]
No ETA at the moment.  We're currently dealing with an influx in customer requests due to 1.0 and then once that settles down I would like us to finish some of the "easy" features like this one.  We did about half the work ~6mo ago and then paused to finish iOS (which we've now done).

**Comment by timi888666** [comment]
Hi @eseidel , Have you started working on this work yet? If not, what are the higher priority tasks?

**Comment by eseidel** [comment]
> Hi @eseidel , Have you started working on this work yet? If not, what are the higher priority tasks?

We've not resumed this work yet (we did some of the needed work a few months ago).  The team is currently focused on finishing iOS support and shipping signed patches.
https://github.com/shorebirdtech/shorebird/issues/1892
https://github.com/shorebirdtech/shorebird/issues/112

Our planning boards are all public: https://github.com/orgs/shorebirdtech/projects/19

**Comment by eseidel** [comment]
As part of doing https://github.com/shorebirdtech/shorebird/issues/112 w'ere back in this code doing more related work (updating the format of the "patch" that we use so that it's a container and can contain more than one file).  We're still probably many weeks away from shipping this though (since we have a bunch of other high priority work to finish first).

**Comment by eseidel** [comment]
We ended up doing #112 without needing to move to a container format (for now), but we've done the planning and so know what we need to do when we get back to this.  We're mostly still focused on finishing https://github.com/shorebirdtech/shorebird/issues/1892 before we take on new features.

**Comment by ronaldoandjames** [comment]
any update?

**Comment by eseidel** [comment]
> any update?

No update at this time.  Felix is currently still assisting existing customers in adding iOS and Bryan is currently working on https://github.com/shorebirdtech/shorebird/issues/2223.

**Comment by alur2191** [comment]
Would love to see this happen 👍

**Comment by rockingdice** [comment]
I'm still waiting for this feature to decide whether to use this service. Since #2223 is done, any plan? Will you make this prioritized?

**Comment by ronaldoandjames** [comment]
me either，  it's really an important feature

rockingdice ***@***.***>于2024年9月7日 周六上午11:30写道：

**Comment by mubashardev** [comment]
## Reminder
I'm still waiting for this feature. Please don't forget it 🥺

**Comment by eseidel** [comment]
We've not forgotten. :) Just busy with other things.  Currently working on shipping https://github.com/shorebirdtech/shorebird/issues/739

**Comment by MiMiOne** [comment]
We hope to support APK as soon as possible,desperately need this feature

**Comment by SamuelGadiel** [comment]
Great feature.
High expectations on this.

**Comment by phamlinhk49kmt01** [comment]
2025 do you have plans to develop it?

**Comment by eseidel** [comment]
> 2025 do you have plans to develop it?

We're currently working on https://github.com/shorebirdtech/shorebird/issues/397, but this is likely to be our next big feature after that.

**Comment by Soliardi-Chyngyz** [comment]
Hello, could you please answer my question, is it possible to deactivate shorebird remotely (say via firebase config) in a project that is already released and running?

It's all about last year's crash on your side (applications were covered with symbols..) after which people had to urgently release a new release. I would like to be on the safe side and in such a case simply disable shorebird and the application would roll back to the default code (of the latest release).

duplicate (from linkedin)

**Comment by felangel** [comment]
> Hello, could you please answer my question, is it possible to deactivate shorebird remotely (say via firebase config) in a project that is already released and running?
> 
> It's all about last year's crash on your side (applications were covered with symbols..) after which people had to urgently release a new release. I would like to be on the safe side and in such a case simply disable shorebird and the application would roll back to the default code (of the latest release).
> 
> duplicate (from linkedin)

You can now rollback patches from the console and the patches will be uninstalled from the devices and the devices will revert to the base release.

https://docs.shorebird.dev/code-push/rollback/

**Comment by Soliardi-Chyngyz** [comment]
> 
> You can now rollback patches from the console and the patches will be uninstalled from the devices and the devices will revert to the base release.
> 
> https://docs.shorebird.dev/code-push/rollback/

I thought about this option, but how reliable is it? Will it definitely roll back to the base release version? Thanks

**Comment by eseidel** [comment]
> > You can now rollback patches from the console and the patches will be uninstalled from the devices and the devices will revert to the base release.
> > https://docs.shorebird.dev/code-push/rollback/
> 
> I thought about this option, but how reliable is it? Will it definitely roll back to the base release version? Thanks

We've not had any complaints of it not working. 🤷  If you have more questions unrelated to assets, we should start a new bug rather than notifying all the individuals watching this bug for asset information.  Thanks!

**Comment by lucasrvimieiro** [comment]
Do we have any updates on this? Our company is really looking forward into this feature for shorebird

**Comment by eseidel** [comment]
We have no updates on this yet, no.  Right now Felix is working on some backend-scaling issues and I'm working on https://github.com/shorebirdtech/shorebird/issues/2965

**Comment by GeylanKalafMohe** [comment]
> We have no updates on this yet, no.  Right now Felix is working on some backend-scaling issues and I'm working on https://github.com/shorebirdtech/shorebird/issues/2965

Quite disappointing, this being one of the most requested features. And you said since one year that you will be focusing on this feature.

This feature is mostly preventing me from using Shorebird actually.

**Comment by hodakl099** [comment]
Any updates on this feature yet?

**Comment by eseidel** [comment]
No updates at this time.

**Comment by Jonny1987** [comment]
Do you have any timeline around when this feature will be ready?

**Comment by hodakl099** [comment]
@eseidel This has been one of the most requested feature. Users of shorebird would greatly appreciate an update, as many users are still eagerly waiting for it.

**Comment by eseidel** [comment]
Agreed that it's oft requested.  We don't currently have a timeline.  All our efforts are focused on simply managing the scale of growth of our existing product at the moment.  I'll be sure to update once we have a schedule.

**Comment by Jinwook-Song** [comment]
@eseidel , It looks like this feature was previously included as a “next big feature” in [1.1](https://github.com/orgs/shorebirdtech/projects/19),
but that project has since been closed. I also noticed it isn’t listed in the current to-do items of [April 2025](https://github.com/orgs/shorebirdtech/projects/22)


Does this mean the feature is no longer planned to be supported? Could you clarify the current status and roadmap for it?

**Comment by melikhanhosdogdu** [comment]
Any updates on this feature yet? 

We keep olur lang files as asset and it's really important to us updating any text on the app.

**Comment by felangel** [comment]
> Any updates on this feature yet?
> 
> We keep olur lang files as asset and it's really important to us updating any text on the app.

No updates yet but I highly recommend considering using the [officially recommended localization approach](https://docs.flutter.dev/ui/accessibility-and-internationalization/internationalization) which would support patching any text in your app.

**Comment by mjl-ops** [comment]
@felangel @eseidel I really suggest you take this serious and consider redirecting some of your resources, increase the size of your team or something. Assets like fonts and images are nothing major requiring a whole new release for a developer to push. I am looking at this thread and it looks like you have been talking about this feature since 2023 and that is ancient. Guys! please get back to work. Imagine im African and I see the importance of this. Guys! Please! Get back to work! Im going back to ride my elephant. Stop disturbing me.

**Comment by iboldurev** [comment]
any news?

**Comment by eseidel** [comment]
No news.  I'd love for us to do this, but it's a non-trivial amount of engineering (that I don't currently have a good plan for us to recoop the costs of), and we need to grow our team first.  But I would like to add this once we have more capacity to do so.

**Comment by gregjkal** [comment]
[The docs say](https://github.com/shorebirdtech/docs/blame/7a5065f6f94862df8ca182a92cd1fe32f9917b7c/src/content/docs/code-push/index.mdx#L86):

> we have plans to support this in the near future

That was written at least two years ago, and the latest status update indicates no clear plans to implement this. I totally understand the capacity constraints. I would suggest changing the language in the docs about "near future". Most people don't think that means "in some number of years" 😊.

**Comment by eseidel** [comment]
Agreed. :) We actually have someone staffed to do this now, although they're currently working on https://github.com/shorebirdtech/shorebird/issues/2965 first.  They built a demo of assets, but we would need to do more work to ship it.

**Comment by techwithsam** [comment]
Hello @eseidel, any update on this?

**Comment by eseidel** [comment]
No update.  We're still focused on https://github.com/shorebirdtech/shorebird/issues/2965 for now.

**Comment by eseidel** [comment]
Just wanted to say this remains high on our list. We've built multiple internal demos over the years, so the runtime side isn't really the blocker -- most of the remaining work is on the hosting and billing side, since assets change what a patch actually is and therefore what we meter. Some of the billing work we shipped recently was groundwork for exactly this, so we're closer than we were. Still no date I can commit to: we're a two-full-time-engineer team, and that's the real constraint. Hiring now, and I'll update here as this moves.
